### PR TITLE
Add restriction to first usage in allocation planner

### PR DESCRIPTION
**Pull request by wangyems**
**Description**: Describe your changes.

As [discussed](https://github.com/microsoft/onnxruntime/pull/10704#discussion_r817313539), fix the bug described [here](https://github.com/microsoft/onnxruntime/pull/10704#issue-1154732465) without changing the "first usage" rule. 

**Motivation and Context**
- Why is this change required? What problem does it solve?
- If it fixes an open issue, please link to the issue here.

Fix crash issue on an 1st party model

### CODE REVIEW

File: onnxruntime/core/framework/allocation_planner.cc [hunk 1]
```diff
@@ -790,7 +795,8 @@ class PlannerImpl {
   Status GeneratePlanForWeights() {
     // TODO: Move away from usage of vector of `OrtMemoryInfo`s per weight (initializer)
     // We do not need to maintain a vector of locations that a weight is used in.
-    // We only need to know the location of its first usage because:
+    // We only need to know the location of its first usage(if it has consumers in main
```
Comment (hariharans29) [inline onnxruntime/core/framework/allocation_planner.cc:798]: Nit: Same comment as above. "Main graph" in this context is a bit inaccurate.

**Comment by hariharans29** [comment]
LGTM. 

It would be nice to craft a toy model test for this. Maybe a model with just 2 nodes: an If node with a subgraph containing a node that consumes an initializer on CUDA followed by any other node in the main graph level that consumes the same initializer on CPU. That should trigger the bug with the old logic and keep this change covered.

@skottmckay: Any comments ?

**Comment by wangyems** [comment]
> LGTM.
> 
> It would be nice to craft a toy model test for this. Maybe a model with just 2 nodes: an If node with a subgraph containing a node that consumes an initializer on CUDA (initializer introduced in the main graph level) followed by any other node in the main graph level that consumes the same initializer on CPU (an improvised version of `LocationPlanningForInitializersOnlyUsedInANestedSubgraph`). That should trigger the bug with the old logic and keep this change covered.
> 
> @skottmckay: Any comments ?

UT added.

**Review by skottmckay — APPROVED** [review]
:shipit:

**Review by hariharans29 — APPROVED** [review]

**Pull request merged**